cms/taxonomy/management/commands/sync_topics.py

```diff
@@ -0,0 +1,203 @@
+        topics.extend(subtopics)
+
+        # Add any subtopics URLs from the topics we just fetched to the stack
+        request_stack.extend(_extract_subtopic_links(raw_topics))
+
+    return topics
+
+
+def _request_topics(url: str) -> list[dict[str, str]]:
+    """Fetch topics from the API and return the items from the response."""
+    topics_response = requests.get(url, timeout=30)
+    topics_response.raise_for_status()
+
+    return topics_response.json().get("items", [])
+
+
+def _extract_subtopic_links(raw_topics: list[dict[str, Any]]) -> list[tuple[str, str]]:
+    """Return a list of tuples of any subtopic links and their parent topic IDs found in raw_topics."""
+    return [
+        (subtopic_link, raw_topic["id"])
+        for raw_topic in raw_topics
+        # NOTE: This assumes we can simply call the links in the response as they are provided,
+        # perhaps switch to building the URL ourselves
+        if (
+            (subtopic_link := raw_topic.get("links", {}).get("subtopics", {}).get("href"))
+            and raw_topic.get("subtopics_ids")
+        )
+    ]
+
+
+def _sync_with_fetched_topics(fetched_topics: list[dict[str, str]]):
+    """For each fetched topic, decide if it needs to be created, updated, or left as is."""
+    _check_for_duplicate_topics(fetched_topics)
+
+    updated_count = 0
+    created_count = 0
+    for fetched_topic in fetched_topics:
+        if existing_topic := _get_topic(fetched_topic["id"]):
+            if not _topic_matches(existing_topic, fetched_topic):
+                _update_topic(existing_topic, fetched_topic)
+                updated_count += 1
+        else:
+            _create_topic(fetched_topic)
+            created_count += 1
+
+    logger.info("Saved %d new topic(s)", created_count)
+    logger.info("Updated %d existing topic(s)", updated_count)
+
+
+def _check_for_duplicate_topics(fetched_topics: list[dict[str, str]]) -> None:
+    topic_ids = set()
+    for topic in fetched_topics:
+        if topic["id"] in topic_ids:
+            raise RuntimeError(f"Received duplicate topic ID in API responses, topic IDs must be unique: {topic['id']}")
+        topic_ids.add(topic["id"])
+
+
+def _get_topic(topic_id: str) -> Topic | None:
+    """Fetches a Topic record by its primary key (id), or returns None if none is found."""
+    return Topic.objects.filter(id=topic_id).first()
+
+
+def _topic_matches(existing_topic: Topic, fetched_topic: dict[str, str]) -> bool:
+    """Compares: title, description, removed status, Parent ID.
+    If all these match, the function returns True; otherwise False.
+    """
+    existing_parent = existing_topic.get_parent()
+    existing_parent_id = existing_parent.id if existing_parent else None
+
+    return (
+        fetched_topic["title"] == existing_topic.title
+        and fetched_topic.get("description") == existing_topic.description
```

Comment:
I'm not convinced these should be checked. I can see corrections being made to the title/description. The title might be less problematic. I think if the topic ID and parent ID match and maybe title, it should be enough.

Thoughts?

Reply:
Ah ofc 👍